File: bronze/lineup.py
def main():
    condx = []; condy = []
    with open('lineup.in', 'r') as f:
        for _ in range(int(f.readline())):
            tmp = f.readline().split()
            condx.append(tmp[0])
            condy.append(tmp[-1])

    cows = ["Bessie", "Buttercup", "Belinda", "Beatrice",
    		"Bella", "Blue", "Betsy", "Sue"]
    cows = sorted(cows)

    res = []
    while len(res) != 8:
        chain = 0
        if len(res):
            for c in cows:
                for j in range(len(condx)):
                    if (condx[j] == c and condy[j] == res[-1]) or \
                        (condy[j] == c and condx[j] == res[-1]):
                        res.append(c)
                        cows.remove(c)
                        chain = 1
                        break
                if (chain):
                    break

        if not chain:
            for c in cows:
                neigh = []
                for j in range(len(condx)):
                    if condx[j] == c and condy[j] not in res:
                        neigh.append(condy[j])
                    elif condy[j] == c and condx[j] not in res:
                        neigh.append(condx[j])
                if len(neigh) < 2:
                    res.append(c)
                    cows.remove(c)
                    break

    with open('lineup.out', 'w') as f:
        for i in res:
            f.write(i+'\n')

if __name__ == '__main__':
    main()

File: bronze/lineup.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
	freopen("lineup.in", "r", stdin);
	freopen("lineup.out", "w", stdout);

	int n; cin >> n;
	vector<pair<string, string>> conds;
	for (int i = 0; i < n; i++) {
		string cow1, tmp, cow2;
		cin >> cow1;
		for (int j = 0; j < 4; j++) cin >> tmp;
		cin >> cow2;
		conds.push_back({cow1, cow2});
	}

	vector<string> cows = {"Bessie", "Buttercup", "Belinda", "Beatrice",
							"Bella", "Blue", "Betsy", "Sue"};
	sort(begin(cows), end(cows));

	do {
		bool valid = true;
		for (auto &c: conds) {
			auto a = find(begin(cows), end(cows), c.first);
			auto b = find(begin(cows), end(cows), c.second);
			if (abs(a-b) > 1) {
				valid = false;
			}
		}
		if (valid) {
			break;
		}
	} while(next_permutation(begin(cows), end(cows)));

	for (auto &x : cows) cout << x << endl;
	return 0;
}
